File: Kernel Simulator/Network/RPC/RPC-Commands.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Net.Sockets

Public Module RPC_Commands

    ''' <summary>
    ''' List of RPC commands.<br/>
    ''' <br/>&lt;Request:Shutdown&gt;: Request will be like this: &lt;Request:Shutdown&gt;(IP)
    ''' <br/>&lt;Request:Reboot&gt;: Request will be like this: &lt;Request:Reboot&gt;(IP)
    ''' <br/>&lt;Request:Lock&gt;: Request will be like this: &lt;Request:Lock&gt;(IP)
    ''' <br/>&lt;Request:SaveScr&gt;: Request will be like this: &lt;Request:SaveScr&gt;(IP)
    ''' <br/>&lt;Request:Exec&gt;: Request will be like this: &lt;Request:Exec&gt;(Lock)
    ''' <br/>&lt;Request:Acknowledge&gt;: Request will be like this: &lt;Request:Acknowledge&gt;(IP)
    ''' <br/>&lt;Request:Ping&gt;: Request will be like this: &lt;Request:Ping&gt;(IP)
    ''' </summary>
    ReadOnly RPCCommands As New List(Of String) From {"<Request:Shutdown>", "<Request:Reboot>", "<Request:Lock>", "<Request:SaveScr>", "<Request:Exec>", "<Request:Acknowledge>", "<Request:Ping>"}

    ''' <summary>
    ''' Send an RPC command to another instance of KS using the specified address
    ''' </summary>
    ''' <param name="Request">A request</param>
    ''' <param name="IP">An IP address which the RPC is hosted</param>
    Public Sub SendCommand(Request As String, IP As String)
        SendCommand(Request, IP, RPCPort)
    End Sub

    ''' <summary>
    ''' Send an RPC command to another instance of KS using the specified address
    ''' </summary>
    ''' <param name="Request">A request</param>
    ''' <param name="IP">An IP address which the RPC is hosted</param>
    ''' <param name="Port">A port which the RPC is hosted</param>
    ''' <exception cref="InvalidOperationException"></exception>
    Public Sub SendCommand(Request As String, IP As String, Port As Integer)
        If RPCEnabled Then
            Dim Cmd As String = Request.Remove(Request.IndexOf("("))
            Wdbg(DebugLevel.I, "Command: {0}", Cmd)
            Dim Arg As String = Request.Substring(Request.IndexOf("(") + 1)
            Wdbg(DebugLevel.I, "Prototype Arg: {0}", Arg)
            Arg = Arg.Remove(Arg.Count - 1)
            Wdbg(DebugLevel.I, "Finished Arg: {0}", Arg)
            Dim Malformed As Boolean
            If RPCCommands.Contains(Cmd) Then
                Wdbg(DebugLevel.I, "Command found.")
                Dim ByteMsg() As Byte = {}
                If Cmd = "<Request:Shutdown>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0}", Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("ShutdownConfirm, " + Arg + vbNewLine)
                ElseIf Cmd = "<Request:Reboot>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0}", Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("RebootConfirm, " + Arg + vbNewLine)
                ElseIf Cmd = "<Request:Lock>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0}", Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("LockConfirm, " + Arg + vbNewLine)
                ElseIf Cmd = "<Request:SaveScr>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0}", Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("SaveScrConfirm, " + Arg + vbNewLine)
                ElseIf Cmd = "<Request:Exec>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0} to execute ""{1}""", IP, Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("ExecConfirm, " + Arg + vbNewLine)
                ElseIf Cmd = "<Request:Acknowledge>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0}", Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("AckConfirm, " + Arg + vbNewLine)
                ElseIf Cmd = "<Request:Ping>" Then
                    Wdbg(DebugLevel.I, "Stream opened for device {0}", Arg)
                    ByteMsg = Text.Encoding.Default.GetBytes("PingConfirm, " + Arg + vbNewLine)
                Else
                    Wdbg(DebugLevel.E, "Malformed request. {0}", Cmd)
                    Malformed = True
                End If
                If Not Malformed Then
                    Wdbg(DebugLevel.I, "Sending response to device...")
                    RPCListen.Send(ByteMsg, ByteMsg.Length, IP, Port)
                    EventManager.RaiseRPCCommandSent(Cmd)
                End If
            End If
        Else
            Throw New InvalidOperationException(DoTranslation("Trying to send an RPC command while RPC didn't start."))
        End If
    End Sub

    ''' <summary>
    ''' Thread to listen to commands.
    ''' </summary>
    Sub RecCommand()
        Dim endp As New IPEndPoint(IPAddress.Any, RPCPort)
        While RPCThread.IsAlive
            Dim buff() As Byte
            Dim ip As String = ""
            Dim msg As String = ""
            Try
                buff = RPCListen.Receive(endp)
                msg = Text.Encoding.Default.GetString(buff)
                Wdbg("RPC: Received message {0}", msg)
                EventManager.RaiseRPCCommandReceived(msg)
                If msg.StartsWith("ShutdownConfirm") Then
                    Wdbg(DebugLevel.I, "Shutdown confirmed from remote access.")
                    RPCPowerListener.Start("shutdown")
                ElseIf msg.StartsWith("RebootConfirm") Then
                    Wdbg(DebugLevel.I, "Reboot confirmed from remote access.")
                    RPCPowerListener.Start("reboot")
                ElseIf msg.StartsWith("LockConfirm") Then
                    Wdbg(DebugLevel.I, "Lock confirmed from remote access.")
                    LockScreen()
                ElseIf msg.StartsWith("SaveScrConfirm") Then
                    Wdbg(DebugLevel.I, "Save screen confirmed from remote access.")
                    ShowSavers(defSaverName)
                ElseIf msg.StartsWith("ExecConfirm") Then
                    If LoggedIn Then
                        Wdbg(DebugLevel.I, "Exec confirmed from remote access.")
                        Console.WriteLine()
                        GetLine(False, msg.Replace("ExecConfirm, ", "").Replace(vbNewLine, ""))
                    Else
                        Wdbg(DebugLevel.W, "Tried to exec from remote access while not logged in. Dropping packet...")
                    End If
                ElseIf msg.StartsWith("AckConfirm") Then
                    Wdbg(DebugLevel.I, "{0} says ""Hello.""", msg.Replace("AckConfirm, ", "").Replace(vbNewLine, ""))
                ElseIf msg.StartsWith("PingConfirm") Then
                    Dim IPAddr As String = msg.Replace("PingConfirm, ", "").Replace(vbNewLine, "")
                    Wdbg(DebugLevel.I, "{0} pinged this device!", IPAddr)
                    NotifySend(New Notification(DoTranslation("Ping!"),
                                                DoTranslation("{0} pinged you.").FormatString(IPAddr),
                                                NotifPriority.Low, NotifType.Normal))
                Else
                    Wdbg(DebugLevel.W, "Not found. Message was {0}", msg)
                End If
            Catch ex As Exception
                Dim SE As SocketException = CType(ex.InnerException, SocketException)
                If SE IsNot Nothing Then
                    If Not SE.SocketErrorCode = SocketError.TimedOut Then
                        Wdbg(DebugLevel.E, "Error from host {0}: {1}", ip, SE.SocketErrorCode.ToString)
                        WStkTrc(ex)
                    End If
                Else
                    Wdbg(DebugLevel.E, "Fatal error: {0}", ex.Message)
                    WStkTrc(ex)
                    EventManager.RaiseRPCCommandError(msg, ex)
                End If
            End Try
        End While
    End Sub

End Module

File: Kernel Simulator/Network/RPC/RPCCommands.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using Extensification.StringExts;
using KS.Kernel;
using KS.Kernel.Debugging;
using KS.Kernel.Power;
using KS.Languages;
using KS.Misc.Notifications;
using KS.Misc.Screensaver;

namespace KS.Network.RPC
{
    public static class RPCCommands
    {

        /// <summary>
        /// List of RPC commands.<br/>
        /// <br/>&lt;Request:Shutdown&gt;: Shuts down the remote kernel. Usage: &lt;Request:Shutdown&gt;(IP)
        /// <br/>&lt;Request:Reboot&gt;: Reboots the remote kernel. Usage: &lt;Request:Reboot&gt;(IP)
        /// <br/>&lt;Request:RebootSafe&gt;: Reboots the remote kernel to safe mode. Usage: &lt;Request:RebootSafe&gt;(IP)
        /// <br/>&lt;Request:Lock&gt;: Locks the computer remotely. Usage: &lt;Request:Lock&gt;(IP)
        /// <br/>&lt;Request:SaveScr&gt;: Saves the screen remotely. Usage: &lt;Request:SaveScr&gt;(IP)
        /// <br/>&lt;Request:Exec&gt;: Executes a command remotely. Usage: &lt;Request:Exec&gt;(Lock)
        /// <br/>&lt;Request:Acknowledge&gt;: Pings the remote kernel silently. Usage: &lt;Request:Acknowledge&gt;(IP)
        /// <br/>&lt;Request:Ping&gt;: Pings the remote kernel with notification. Usage: &lt;Request:Ping&gt;(IP)
        /// </summary>
        private readonly static List<string> RPCCommandsField = new()
        {
            "<Request:Shutdown>",
            "<Request:Reboot>",
            "<Request:RebootSafe>",
            "<Request:Lock>",
            "<Request:SaveScr>",
            "<Request:Exec>",
            "<Request:Acknowledge>",
            "<Request:Ping>"
        };

        /// <summary>
        /// Send an RPC command to another instance of KS using the specified address
        /// </summary>
        /// <param name="Request">A request</param>
        /// <param name="IP">An IP address which the RPC is hosted</param>
        public static void SendCommand(string Request, string IP)
        {
            SendCommand(Request, IP, RemoteProcedure.RPCPort);
        }

        /// <summary>
        /// Send an RPC command to another instance of KS using the specified address
        /// </summary>
        /// <param name="Request">A request</param>
        /// <param name="IP">An IP address which the RPC is hosted</param>
        /// <param name="Port">A port which the RPC is hosted</param>
        /// <exception cref="InvalidOperationException"></exception>
        public static void SendCommand(string Request, string IP, int Port)
        {
            if (RemoteProcedure.RPCEnabled)
            {
                // Get the command and the argument
                string Cmd = Request.Remove(Request.IndexOf("("));
                DebugWriter.WriteDebug(DebugLevel.I, "Command: {0}", Cmd);
                string Arg = Request.Substring(Request.IndexOf("(") + 1);
                DebugWriter.WriteDebug(DebugLevel.I, "Prototype Arg: {0}", Arg);
                Arg = Arg.Remove(Arg.Length - 1);
                DebugWriter.WriteDebug(DebugLevel.I, "Finished Arg: {0}", Arg);

                // Check the command
                var Malformed = default(bool);
                if (RPCCommandsField.Contains(Cmd))
                {
                    DebugWriter.WriteDebug(DebugLevel.I, "Command found.");

                    // Check the request type
                    string RequestType = Cmd.Substring(Cmd.IndexOf(":") + 1, Finish: Cmd.IndexOf(">"));
                    var ByteMsg = Array.Empty<byte>();
                    switch (RequestType ?? "")
                    {
                        case "Shutdown":
                        case "Reboot":
                        case "RebootSafe":
                        case "Lock":
                        case "SaveScr":
                        case "Exec":
                        case "Acknowledge":
                        case "Ping":
                            {
                                // Populate the byte message to send the confirmation to
                                DebugWriter.WriteDebug(DebugLevel.I, "Stream opened for device {0}", Arg);
                                ByteMsg = System.Text.Encoding.Default.GetBytes($"{RequestType}Confirm, " + Arg + Kernel.Kernel.NewLine);
                                break;
                            }

                        default:
                            {
                                // Rare case reached. Drop it.
                                DebugWriter.WriteDebug(DebugLevel.E, "Malformed request. {0}", Cmd);
                                Malformed = true;
                                break;
                            }
                    }

                    // Send the response
                    if (!Malformed)
                    {
                        DebugWriter.WriteDebug(DebugLevel.I, "Sending response to device...");
                        RemoteProcedure.RPCListen.Send(ByteMsg, ByteMsg.Length, IP, Port);
                        Kernel.Kernel.KernelEventManager.RaiseRPCCommandSent(Cmd, Arg, IP, Port);
                    }
                }
            }
            else
            {
                throw new InvalidOperationException(Translate.DoTranslation("Trying to send an RPC command while RPC didn't start."));
            }
        }

        /// <summary>
        /// Thread to listen to commands.
        /// </summary>
        public static void ReceiveCommand()
        {
            var RemoteEndpoint = new IPEndPoint(IPAddress.Any, RemoteProcedure.RPCPort);
            while (RemoteProcedure.RPCStarted)
            {
                Thread.Sleep(1);
                byte[] MessageBuffer;
                string Message = "";
                try
                {
                    MessageBuffer = RemoteProcedure.RPCListen.Receive(ref RemoteEndpoint);
                    Message = System.Text.Encoding.Default.GetString(MessageBuffer);

                    // If the message is not empty, parse it
                    if (!string.IsNullOrEmpty(Message))
                    {
                        DebugWriter.WriteDebug((DebugLevel)Convert.ToInt32("RPC: Received message {0}"), Message);
                        Kernel.Kernel.KernelEventManager.RaiseRPCCommandReceived(Message, RemoteEndpoint.Address.ToString(), RemoteEndpoint.Port);

                        // Iterate through every confirmation message
                        if (Message.StartsWith("ShutdownConfirm"))
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Shutdown confirmed from remote access.");
                            KernelTools.RPCPowerListener.Start(PowerMode.Shutdown);
                        }
                        else if (Message.StartsWith("RebootConfirm"))
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Reboot confirmed from remote access.");
                            KernelTools.RPCPowerListener.Start(PowerMode.Reboot);
                        }
                        else if (Message.StartsWith("RebootSafeConfirm"))
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Reboot to safe mode confirmed from remote access.");
                            KernelTools.RPCPowerListener.Start(PowerMode.RebootSafe);
                        }
                        else if (Message.StartsWith("LockConfirm"))
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Lock confirmed from remote access.");
                            Screensaver.LockScreen();
                        }
                        else if (Message.StartsWith("SaveScrConfirm"))
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Save screen confirmed from remote access.");
                            Screensaver.ShowSavers(Screensaver.DefSaverName);
                        }
                        else if (Message.StartsWith("ExecConfirm"))
                        {
                            if (Flags.LoggedIn)
                            {
                                DebugWriter.WriteDebug(DebugLevel.I, "Exec confirmed from remote access.");
                                ConsoleBase.ConsoleWrapper.WriteLine();
                                Shell.Shell.GetLine(Message.Replace("ExecConfirm, ", "").Replace(Kernel.Kernel.NewLine, ""));
                            }
                            else
                            {
                                DebugWriter.WriteDebug(DebugLevel.W, "Tried to exec from remote access while not logged in. Dropping packet...");
                            }
                        }
                        else if (Message.StartsWith("AckConfirm"))
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "{0} says \"Hello.\"", Message.Replace("AckConfirm, ", "").Replace(Kernel.Kernel.NewLine, ""));
                        }
                        else if (Message.StartsWith("PingConfirm"))
                        {
                            string IPAddr = Message.Replace("PingConfirm, ", "").Replace(Kernel.Kernel.NewLine, "");
                            DebugWriter.WriteDebug(DebugLevel.I, "{0} pinged this device!", IPAddr);
                            Notifications.NotifySend(new Notification(Translate.DoTranslation("Ping!"), Translate.DoTranslation("{0} pinged you.").FormatString(IPAddr), Notifications.NotifPriority.Low, Notifications.NotifType.Normal));
                        }
                        else
                        {
                            DebugWriter.WriteDebug(DebugLevel.W, "Not found. Message was {0}", Message);
                        }
                    }
                }
                catch (Exception ex)
                {
                    SocketException SE = (SocketException)ex.InnerException;
                    if (SE is not null)
                    {
                        if (!(SE.SocketErrorCode == SocketError.TimedOut))
                        {
                            DebugWriter.WriteDebug(DebugLevel.E, "Error from host: {0}", SE.SocketErrorCode.ToString());
                            DebugWriter.WriteDebugStackTrace(ex);
                        }
                    }
                    else
                    {
                        DebugWriter.WriteDebug(DebugLevel.E, "Fatal error: {0}", ex.Message);
                        DebugWriter.WriteDebugStackTrace(ex);
                        Kernel.Kernel.KernelEventManager.RaiseRPCCommandError(Message, ex, RemoteEndpoint.Address.ToString(), RemoteEndpoint.Port);
                    }
                }
            }
        }

    }
}
